File: src/main/java/nnglebanov/daoexample/services/business/ObjectsPrintService.java
package nnglebanov.daoexample.services.business;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import nnglebanov.daoexample.repositories.BookRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class ObjectsPrintService {
    @NonNull
    private final BookRepository bookRepository;
    private final String DELIMITER = "-----------------";

    @Transactional
    public void outAllBooks() {
        System.out.println(DELIMITER);
        bookRepository.findAll().forEach(
                x -> System.out.println(
                        "Book number: " + x.getId() + "\n" +
                                "Book title: " + x.getBookTitle() + "\n" +
                                "Book authors: " + x.getAuthors().stream()
                                .map(y -> y.getFirstName() + " " + y.getLastName())
                                .collect(Collectors.joining(","))
                                + "\n" + DELIMITER
                )
        );
    }
}

File: src/main/java/nnglebanov/daoexample/services/business/ObjectsPrintService.kt
package nnglebanov.daoexample.services.business

import nnglebanov.daoexample.repositories.BookRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import java.util.stream.Collectors

@Service
class ObjectsPrintService(val bookRepository: BookRepository) {

    private val DELIMITER = "-----------------"
    @Transactional
    fun outAllBooks() {
        println(DELIMITER)
        bookRepository.findAll().subscribe {
            println("Book number: " + it.id + "\n" +
                    "Book title: " + it.bookTitle + "\n" +
                    "Book authors: " + it.authors?.stream()
                    ?.map { y -> y.firstName + " " + y.lastName }
                    ?.collect(Collectors.joining(","))
            + "\n" + DELIMITER)
        }
    }
}
